File: day-01/go/code.go
package main

import (
  "bufio"
  "fmt"
  "log"
  "os"
  "strconv"
)

func main() {
  args := os.Args

  file, err := os.Open(args[1])
  if err != nil { log.Fatal(err) }
  defer file.Close()

  lines := bufio.NewScanner(file)
  total_fuel := 0
  for lines.Scan() {
    module_mass, err := strconv.Atoi(lines.Text())
    if err != nil { log.Fatal(err) }

    fuel := module_mass / 3 - 2
    fuel_for_fuel := fuel
    for {
      fuel_for_fuel = fuel_for_fuel / 3 - 2
      if fuel_for_fuel <= 0 {
        break
      }
      fuel += fuel_for_fuel
    }

    fmt.Println(module_mass, "->", fuel)
    total_fuel += fuel
  }

  fmt.Println("Total fuel:", total_fuel)
}

File: day-01/rust/code.rs
use std::env;
use std::io::BufReader;
use std::io::BufRead;
use std::fs::File;

fn eval_module_mass(module_mass: i32) -> i32 {
  let mut fuel = module_mass / 3 - 2;
  let mut fuel_for_fuel = fuel;
  loop {
    fuel_for_fuel = fuel_for_fuel / 3 - 2;
    if fuel_for_fuel < 0 {
      break;
    }
    fuel += fuel_for_fuel;
  }
  return fuel;
}

fn main() {
  let args: Vec<String> = env::args().collect();

  let f = File::open(&args[1]).unwrap();
  let file = BufReader::new(&f);

  let mut total_fuel: i32 = 0;

  for line in file.lines() {
    let module_mass: i32 = line.unwrap().parse::<i32>().unwrap();
    let this_module_fuel: i32 = eval_module_mass(module_mass);
    //println!("{} -> {}", module_mass, this_module_fuel);
    total_fuel += this_module_fuel;
  }

  println!("Total fuel required: {}", total_fuel);
}

